core/src/test/scala/integration/kafka/api/TransactionsTest.scala

```diff
@@ -814,11 +913,31 @@ class TransactionsTest extends IntegrationTestHarness {
     producer3.send(TestUtils.producerRecordWithExpectedTransactionStatus(topic1, 0, "4", "4", willBeCommitted = true))
     producer3.commitTransaction()
 
-    // Check that the epoch only increased by 1
     producerStateEntry =
       brokers(partitionLeader).logManager.getLog(new TopicPartition(topic1, 0)).get.producerStateManager.activeProducers.get(producerId)
     assertNotNull(producerStateEntry)
-    assertEquals((initialProducerEpoch + 1).toShort, producerStateEntry.producerEpoch)
+
+    // Check that the epoch only increased by 1 when TV2 is disabled.
+    // With TV2 and the latest EndTxnRequest version, the epoch will be bumped at the end of every transaction.
+    if (!isTV2Enabled) assertEquals((initialProducerEpoch + 1).toShort, producerStateEntry.producerEpoch)
+    else {
+      // Producer State entry contains the last epoch with which records were sent.
+      assertEquals((initialProducerEpoch + 2).toShort, producerStateEntry.producerEpoch)
+
+      // Access the client's producer epoch via reflection to verify epoch bump on the last End Txn Request.
+      val transactionManagerField: Field = classOf[KafkaProducer[_, _]].getDeclaredField("transactionManager")
```

Comment:
I wanted to make sure that the epoch was exactly 3 on the client after the 3 endTxnReq since it doesn't update on the producerStateEntry until the next commit right? But if it's not necessary I can just check the producer state entry!

Reply:
It should be updated on the commit, but you may need to wait until the marker is written.